### ISSUE TITLE
Router continuously fires `routeChangeError` after using `push` to a dynamic route segment

### ISSUE THREAD

**Issue by omarestrella** [vercel/next.js#48839]
### Verify canary release

- [X] I verified that the issue exists in the latest Next.js canary release

### Provide environment information

```bash
Local environment:
    Operating System:
      Platform: darwin
      Arch: arm64
      Version: Darwin Kernel Version 22.4.0: Mon Mar  6 20:59:28 PST 2023; root:xnu-8796.101.5~3/RELEASE_ARM64_T6000
    Binaries:
      Node: 16.17.0
      npm: 8.15.0
      Yarn: 1.22.17
      pnpm: 7.30.5
    Relevant packages:
      next: 13.2.0
      eslint-config-next: 13.3.0
      react: 18.2.0
      react-dom: 18.2.0

Stackblitz (small recreate):
    Operating System:
      Platform: linux
      Arch: x64
      Version: Ubuntu 20.04.0 LTS Tue Apr 25 2023 15:55:15 GMT-0400 (Eastern Daylight Time)
    Binaries:
      Node: 16.14.2
      npm: 7.17.0
      Yarn: 1.22.19
      pnpm: 8.2.0
    Relevant packages:
      next: 13.2.4
      eslint-config-next: N/A
      react: 18.2.0
      react-dom: 18.2.0
```

### Which area(s) of Next.js are affected? (leave empty if unsure)

Routing (next/router, next/navigation, next/link)

### Link to the code that reproduces this issue

https://stackblitz.com/edit/nextjs-btjwni

### To Reproduce

0. You should open your browser console to see the event handlers being called
1. On the index page, click on the link "Go to /page/[dynamic] with a Router.push"
2. After landing on the dynamic segment page, click on the link "To Random Dynamic With Router Push »"
3. Wait a moment (a few seconds) and click on the link "To Random Dynamic With Router Push »" and you will see the router error event, `routeChangeError`, has fired
4. Continue to use the same link and see that the event continues to fire each time

### Describe the Bug

When you use `Router.push` in succession on a dynamic route segment, the Router continues to fire the `routeChangeError` event. This does not happen if you use a `<Link />` to perform the page transition.

### Expected Behavior

No router errors occur on Router.push calls, like <Link />

### Which browser are you using? (if relevant)

Chrome 112.0.5615.137

**Comment by icyJoseph** [comment]
This is a funky bug you have created yourself, I think anyway.

There's two preconditions:

- Shallow routing enabled
- Pushing onto the same route

If you do those two, even with `next/link`, you'll awaken the route cancellation, because, you are navigating onto the same page, shallowly, so nothing happens.

Try with this:

```jsx
import React from 'react';
import { useRouter } from 'next/router';
import Link from 'next/link';

let inc = 0;
const getInc = () => {
  return inc++;
};

export default function Dynamic() {
  const Router = useRouter();

  return (
    <div>
      <div style={{ marginBottom: '20px' }}>
        On a dynamic route: {Router.query.dynamic}
      </div>

      <div>
        <Link href="/">&laquo; Back to Index</Link>
        <Link href={`/page/${getInc()}`} shallow>
          Go to Random Dynamic with anchor
        </Link>
        <br />
        <a
          href="#"
          onClick={(e) => {
            e.preventDefault();
            Router.push(
              {
                pathname: '/page/' + getInc(),
              },
              undefined,
              { shallow: false }
            );
          }}
        >
          To Random Dynamic With Router Push &raquo;
        </a>
      </div>
    </div>
  );
}
```

In your example, since you use Math.random, the query changes even when cancelled is triggered, obscuring the bug. Remember that rendering should be pure! Math.random breaks that, and while it is not the reason for the bug, again, it makes it harder to see.

I think the one funky thing is that Link will just straight up trigger the route cancellation now, but again, it is because you are navigating into the same path, through push, shallowly.

**Comment by icyJoseph** [comment]
I had initially thought this was because:

> This is a funky bug you have created yourself, I think anyway.
> 
> There's two preconditions:
> 
> - Shallow routing enabled
> - Pushing onto the same route
> 
> If you do those two, even with `next/link`, you'll awaken the route cancellation, because, you are navigating onto the same page, shallowly, so nothing happens.

And was pretty confident about it because of the usage of Math.random.

However, the bug seems to just rooted on shallow routing. All but the first shallow navigations trigger cancelation. Still investigating why.

That is to say that, `next/link`, with shallow also behaves as you describe.

**Comment by icyJoseph** [comment]
I think the error is somehow getting triggered by this part: https://github.com/vercel/next.js/blob/canary/packages/next/src/shared/lib/router/router.ts#L1265-L1276

Not sure about those flags just yet - also wondering if this is really an issue at this point ~

**Comment by omarestrella** [comment]
> I think the error is somehow getting triggered by this part: https://github.com/vercel/next.js/blob/canary/packages/next/src/shared/lib/router/router.ts#L1265-L1276

Yeah, `this.clc` never seems to clear in the expected way. Sure, its reset in this block, but then this runs: https://github.com/vercel/next.js/blob/canary/packages/next/src/shared/lib/router/router.ts#L1960 and resets the clc function on the router.

> Not sure about those flags just yet - also wondering if this is really an issue at this point ~

It's a bit concerning because it makes the error event a bit pointless, or at least not useful for responding/gathering metrics. Is there actually a route error occurring? It doesn't seem that way, the transition to the page works. There is no actual cancellation either, correct? Ultimately, I just can't trust _any_ error the router emits at all at the moment.

**Comment by icyJoseph** [comment]
You can, just check `err.cancelled`, those are safe to ignore.

**Comment by omarestrella** [comment]
To add to all of this, it does seem like quick shallow route changes result the the Router to stop firing the `routeChangeComplete` events as well. So doing something like:
```
router.push(`/?counter=${Math.random()}`, undefined, { shallow: true }).then(() => {
  return router.push(`/?counter=${Math.random()}`, undefined, { shallow: true })
}).then(() => {
  return router.push(`/?counter=${Math.random()}`, undefined, { shallow: true })
})
```
will cause all following shallow routes to fail to fire the `routeChangeComplete` event even though the route change does occur.

Which can occur in user-space if your users are quickly navigating via controls that call `Router.push` (quickly paging through a table that changes a query parameter, for example). I can wire up another example, if necessary. This does seem like a deeper issue with shallow routing on the same (static) pages.
